src/Sarif/Writers/SarifLogger.cs

```diff
@@ -241,14 +243,24 @@ public void AnalysisStopped(RuntimeConditions runtimeConditions)
 
         public void Log(IRule rule, Result result)
         {
+            if (rule.Id != result.RuleId)
+            {
+                //The rule id '{0}' specified by the result does not match the actual id of the rule '{1}'
+                string message = string.Format(SdkResources.ResultRuleIdDoesNotMatchRule,
+                    result.RuleId.ToString(),
+                    rule.Id.ToString());
+
+                throw new ArgumentException(message);
+            }
+
             if (!ShouldLog(result.Level))
             {
                 return;
             }
 
             if (rule != null)
             {
-                Rules[rule.Id] = rule;
+                Rules[result.RuleKey ?? result.RuleId] = rule;
```

Comment:
I could throw here or the logger could automatically disambiguate these items (would require mutating the passed in result as well). preference? (I will do this in a future change, btw)